### ISSUE TITLE
Enhance the EntityLinks API

### ISSUE THREAD

**Issue by jstano** [spring-projects/spring-hateoas#244]
Add additional methods that take varargs for controller parameters for cases where the controller's @RequestMapping annotation contains path variables.

This will simplify the usage of the API in these cases, by not requiring the use of the more verbose "linkFor().slash(id)" syntax.

I've created a PullRequest that includes these changes along with a unit test to show how the changes will work.

**Comment by meyertee** [comment]
I agree that it's problematic not to support nested resources. The `EntityLinks` approach looks great when starting, but becomes unusable pretty quickly in my opinion. That's a pitty and forces devs to refactor their code. Maybe it could be made clearer in the documentation that it's only for the simplest/flat structures.

I guess it could be argued that even without knowledge of the controller, it is domain knowledge that for example to process an `Order` we need to know the customerId.
I wouldn't mind a signature like `EntityLinks.linkForSingleResource(Class<?> type, Object id, Map<String, Object> parameters)`.
It's also not possible to generate templated Links with the simple EntityLinks methods.

**Comment by behrangsa** [comment]
> Generally speaking, yes. EntityLinks is supposed to be used like: "here's the type, give me the collection link" or "here's the object, give me the link to it"

@olivergierke as mentioned by @dsanders1234 and @meyertee, it would be nice if the `EntityLinks`  and `ExposesResourceFor` classes supported nested resources.

For example, by allowing `ExposesResourceFor` on controller methods similar to the following code snippet:

```java
public class CustomerController {

    @GetMapping("/{orderId}")
    @ExposesResourceFor(Order.class);
    public HttpEntity<CustomerResource> getOrder(final @PathVariable String orderId) {
        return null;
    }

    @GetMapping("/{orderId}/line-item/{lineItemId}")
    @ExposesResourceFor(Order.class, LineItem.class);
    public HttpEntity<LineItemResource> getLineItem(final @PathVariable String orderId, final @PathVariable String lineItemId) {
        return null;
    }
```
